### PR TITLE
[flang][AddAliasTags] Generalise tbaa tags beyond functions

### PR DESCRIPTION
**Pull request by tblah**
TL;DR without using the undocumented option this should be NFC. Even with the option, it is probably still NFC in practice.

This patch is intended to upstream some experiment code that might be more broadly useful.

This changes the per-function TBAA trees to be per-(top-level)-symbol. This allows running the pass on other top level operations e.g. openmp reductions. Symbol names should be unique so the trees should not collide (without multiple inlining like in #89829).

The current alias code can't generate useful alias tags for reductions (it just sees the reduction variables come from a block argument and correctly concludes it doesn't know anything about them). But in the future this could be useful to allow array reductions to vectorize. For this reason, I don't think the current code will generate alias tags for existing top level operations other than functions, but I would prefer to leave the option off until there is a real use case just to be safe.

I have tested with the option enabled and do not know of any misscompilations.

### CODE REVIEW

**Comment by tblah** [comment]
Ping for review

**Comment by vzakhari** [comment]
Hi Tom, it looks okay to me, but I do not really see what useful TBAA information we can generate for standalone isolated-from-above OpenMP operations.  Can you please give some examples of how this may evolve and become useful?

It seems to me that we can only attach some useful TBAA tags after such operations are inlined into their users' contexts, but then this change does not really apply to this case.

**Comment by tblah** [comment]
> Hi Tom, it looks okay to me, but I do not really see what useful TBAA information we can generate for standalone isolated-from-above OpenMP operations. Can you please give some examples of how this may evolve and become useful?
> 
> It seems to me that we can only attach some useful TBAA tags after such operations are inlined into their users' contexts, but then this change does not really apply to this case.

Hi, thanks for taking a look.

Here is an example reduction declaration for a fixed size array:
```
omp.declare_reduction @add_reduction_byref_box_3xi32 : !fir.ref<!fir.box<!fir.array<3xi32>>> init {
  ^bb0(%arg0: !fir.ref<!fir.box<!fir.array<3xi32>>>):
    // Allocate and initialize a thread private copy of the reduction variable
    // [...]
    omp.yield(%1 : !fir.ref<!fir.box<!fir.array<3xi32>>>)
  } combiner {
  ^bb0(%arg0: !fir.ref<!fir.box<!fir.array<3xi32>>>, %arg1: !fir.ref<!fir.box<!fir.array<3xi32>>>):
    %0 = fir.load %arg0 : !fir.ref<!fir.box<!fir.array<3xi32>>>
    %1 = fir.load %arg1 : !fir.ref<!fir.box<!fir.array<3xi32>>>
    %c0 = arith.constant 0 : index
    %2:3 = fir.box_dims %0, %c0 : (!fir.box<!fir.array<3xi32>>, index) -> (index, index, index)
    %3 = fir.shape_shift %2#0, %2#1 : (index, index) -> !fir.shapeshift<1>
    %c1 = arith.constant 1 : index
    fir.do_loop %arg2 = %c1 to %2#1 step %c1 unordered {
      %4 = fir.array_coor %0(%3) %arg2 : (!fir.box<!fir.array<3xi32>>, !fir.shapeshift<1>, index) -> !fir.ref<i32>
      %5 = fir.array_coor %1(%3) %arg2 : (!fir.box<!fir.array<3xi32>>, !fir.shapeshift<1>, index) -> !fir.ref<i32>
      %6 = fir.load %4 : !fir.ref<i32>
      %7 = fir.load %5 : !fir.ref<i32>
      %8 = arith.addi %6, %7 : i32
      fir.store %8 to %4 : !fir.ref<i32>
    }
    omp.yield(%arg0 : !fir.ref<!fir.box<!fir.array<3xi32>>>)
  }  cleanup {
   // clean up allocations in the init region
   // [...]
  }
```

It would be great if that do-loop could be vectorized for large arrays. Alias analysis could help with this.

You are right that one approach would be to do all of the inlining in MLIR and add TBAA tags afterwards. This is not how OpenMP currently works. There was a desire to share as much OpenMP codegen with clang as possible so these are only inlined (and outlined) when the MLIR LLVM+OpenMP dialects are transformed into LLVM IR.

Given this design, I would prefer to adapt TBAA tags to understand openmp operations before they are inlined. In this case, the two block arguments do not alias (fir::AliasAnalysis doesn't yet know this and I have not added that to this PR).

**Comment by tblah** [comment]
I have no immediate plans to work on what I described above (until we have a concrete usecase for this optimization). I wanted to upstream this code in case it was useful for anyone else working on the same thing, but I don't mind if you would prefer I abandon this PR until it is actually useful for something.

**Comment by vzakhari** [comment]
Thank you for the explanation, Tom.  Let me review your example in more detail today, and give you more feedback.  I am on a fence about merging this right now :)

**Comment by vzakhari** [comment]
Tom, regarding your example, while I understand the desire to provide better aliasing information for the OpenMP isolated from above operations, I do not think this is related to FIR (meaning Fortran) aliasing rules.  And, in general, I do not think FIR AliasAnalysis can apply Fortran aliasing rules just based on the fact that the memory references originate from the block arguments of some isolated from above operation.  Just think of a pass that, for whatever reason, outlines a set of FIR operations into a standalone function, and some live-in memrefs become block arguments of the function: at this point there is no guarantee that this function follows Fortran aliasing rules for these blocks arguments, since they do not represent dummy arguments in Fortran terms.

I think it is more appropriate to rely on the Fortran rules only when it is known that a source of a memory reference comes from an instantiation of a declaration of a Fortran dummy argument, which is conveyed in via `fir.declare` with a `dummy_scope` operand.

So maybe it is better to actually generate such `fir.declare` operations for the block arguments of the reduction operations, if we can guarantee that the two input memory references never alias.  These `fir.declare` will not help right away, because without `getInstantiationPoint` (from #92472) set to true we will still pass through the `fir.declare`'s.  I think `getSource` run from `fir::AliasAnalysis::alias` should actually collect sets of instantiation points for each memory reference, which will be multiple `fir.declare` operations in the use-def chain.  Then the two sets for the two memory references may be checked element per element to see if any pair of `fir.declare`'s comes from the same function instantiation scope - then the best aliasing rule might be applied across such pairs.  To make this work within `fir::AliasAnalysis` we will need some function scope attribute on all `fir.declare`'s: right now, for example, `fir.declare` for a module variable's instantiation does not have any scope attribute and I have to use the dominance information to find the scope - I think it will be unacceptable to use the dominance information in `fir::AliasAnalysis::alias`, so having the scoping information attached to all `fir.declare` will allow to avoid this.

With all that said, I would advise not merging this.

**Comment by agozillon** [comment]
Not really an area I am intimately familiar with, so I can't quite comment one way or the other, but I wouldn't be against gathering alias information for OpenMP operations, however, I'll leave the discussion to people more familiar with this area/work :-)

**Comment by tblah** [comment]
Thank you for your thoughts Slava. I will close this pull request.
